### DISCUSSION TITLE
Inquiry on Multi-Tenancy Schema Creation and Migration Using Piccolo

### DISCUSSION THREAD

**Discussion by Kokoserver** [piccolo-orm/piccolo#920] [Q&A]
I am currently engaged in the development of a SaaS application that employs a multi-tenant architecture with PostgreSQL as the database system. In this setup, each tenant is allocated a unique schema that contains all the necessary tables specific to their data.

Upon reviewing the available documentation, I have identified a section that describes the modification of the schema within the table definition by passing the schema as an argument to the table class. However, I seek further clarification on the following points:

Is there a provision within the piccolo to automatically generate a separate schema for each new tenant upon registration?
Can the Piccolo migration tools be utilized to facilitate the migration of the database schema for each tenant?
Your assistance is pivotal to the progress of my project, and I would greatly appreciate any guidance or resources you could provide on these matters.

Thank you for your attention to this inquiry.

**Comment by dantownsend** [comment]
It sounds like for each user you want to create a new schema, with the identical database tables.

While Piccolo does allow you to specify which schema a table belongs to:

```python
class MyTable(Table, schema="schema_1"):
    ...
```

There isn't a way to tell the database migrations to create tables in multiple schemas simultaneously.

Maybe you can create a database per client instead.

You could then have a script like this (specifying [which database credentials to use](https://piccolo-orm.readthedocs.io/en/latest/piccolo/engines/index.html?highlight=piccolo_conf#piccolo-conf-environment-variable) each time):

```
export PICCOLO_CONF=user_1
piccolo migrations forwards all

export PICCOLO_CONF=user_2
piccolo migrations forwards all
```

It's probably not what you're after though. I'll have a think if there's a better solution. There are probably some creative solutions which will make it work.

**Reply by Kokoserver** [reply]
Database per client will be overkill for the application and it won't be cost effective on the project side.
I'm thinking of something:

if i have a contextvar  that hold the schema name:


from contextvars import ContextVar

SCHEMA_NAME = ContextVar("schema_name", default="public_schema")

CURRENT_SCHEMA = SCHEMA_NAME.get()


# Define your base table which will be used by all tenants
class TableExample(Table, schema=CURRENT_SCHEMA):
    name = Varchar()
    age = Integer()


from piccolo.apps.migrations.commands.forwards import run_forwards
from piccolo.apps.migrations.commands.new import new


async def run_migrations(schema_list: str):
    for schema in schema_list:
        SCHEMA_NAME.set(schema)
    await new(app_name="my_app", auto=True)
    await run_forwards(app_name="my_app")


something like this

**Reply by dantownsend** [reply]
That's an interesting idea - I didn't consider a contextvar. I haven't tried it, but it could work.

**Reply by Kokoserver** [reply]
My next questions will be how to create new schema with piccolo.
Does running:
await new(app_name="my_app", auto=True)
await run_forwards(app_name="my_app")
run the migration?, and if the schema(from the table) point the migration to the write schema

**Reply by dantownsend** [reply]
The first line will create a migration, and the second line will run it.

```python
await new(app_name="my_app", auto=True)
await run_forwards(app_name="my_app")
```

If you look at the contents of the migration though, it hardcodes which schema a table will be written to. So making it dynamic will be tricky.

You said about a database per client being cost prohibitive, but you can setup many databases on a single database server. 

Most of the projects I work on with Piccolo have a single database shared by all users, you just have to be really careful that each endpoint only returns data belonging to the current user.

**Reply by Kokoserver** [reply]
Thanks, like you said making it dynamic in that sense will be tricky, since the schema will be hardcoded, but i'm thinking if a simple decorator that fetch all the schema from the database can be written then attached to the generated migration forwared and new, which will loop through all the available schema and dynamically change that.

```python 
import functools
from typing import Callable
from typeguard import typechecked
from . import TableExamle


@typechecked
async def for_each_tenant_schema(func: Callable) -> Callable:
    @functools.wraps(func)
    def wrapped():
        schemas = await TableExamle.select(TableExample.schema_name).all()
        if schemas:
         for (schema,) in schemas:
            func(schema)
        else: func("default_schema")
    return wrapped
```

**Reply by dantownsend** [reply]
OK, I think I may have a solution.

When you specify your `DB` you can set the search path, which means it will look in that schema first for tables:

```python
DB = PostgresEngine({
    'database': 'my_database',
    'server_settings': {
        'search_path': 'my_schema',
    }
})
```

You can specify 'extra_nodes' (see [docs](https://piccolo-orm.readthedocs.io/en/latest/piccolo/engines/postgres_engine.html?highlight=nodes#source), so you could add one per schema):

```python
DB = PostgresEngine({
    'database': 'my_database',
    'extra_nodes': {
        'customer_1': PostgresEngine({
            'database': 'my_database',
            'server_settings': {
                'search_path': 'customer_1_schema',
            },
        }),
        'customer_2': PostgresEngine({
            'database': 'my_database',
            'server_settings': {
                'search_path': 'customer_2_schema',
            },
        }),
        'customer_3': PostgresEngine({
            'database': 'my_database',
            'server_settings': {
                'search_path': 'customer_3_schema',
            },
        })
    }
})
```

Whenever you run a query, you can specify which node to run it on. For example:

```python
await MyTable.select().run(node='customer_3')
```

For each of your tables, don't specify the schema arg:

```python
class MyTable(Table):
    ...
```

For the migrations, create them as usual. You'll need a special wrapper script which runs them against all schemas though.

It feels a bit like a hack, but it will do something like:

```python
import asyncio

from piccolo.apps.migrations.commands.forwards import run_forwards
from piccolo.table import TABLE_REGISTRY
from piccolo_conf import DB

async def run():
    for node in DB.extra_nodes:
        DB.config = node.config        
        await run_forwards("all")

if __name__ == '__main__':
    asyncio.run(run)
```

I haven't had a chance to try this - but in theory it should all work 🤞

**Reply by Kokoserver** [reply]
Wow this make sense thanks... I will try this out and give you the feedback

**Reply by Kokoserver** [reply]
Is there functionality for creating schema. 

```python 

New_schema = create_schema("NEWSCHEMA")
```

**Reply by dantownsend** [reply]
Yeah, check out [SchemaManager](https://piccolo-orm.readthedocs.io/en/latest/piccolo/api_reference/index.html#schemamanager).

```python
from piccolo.schema import SchemaManager

>>> await SchemaManager().create_schema(schema_name="my_schema")
```

**Reply by Kokoserver** [reply]
Thanks

**Comment by Kokoserver** [comment]
hello @dantownsend , i keep getting ```partially initialized module 'piccolo_conf' has no attribute 'APP_REGISTRY' (most likely due to a circular import)```
anywhere i use piccolo LazyTableReference because i have lot of apps, is there a better way to handle this?.

According to the example on the documentation on many to many both, the joining key must be added to both table that the joining key reference.

**Reply by dantownsend** [reply]
Try doing the import inside the function body.

**Reply by Kokoserver** [reply]
you mean, something like this:
```python 
from app1.tables import Table1
from app2.tables import Table2
```python APP_CONFIG = AppConfig(
    app_name="tenant",
    migrations_folder_path=os.path.join(CURRENT_DIRECTORY, "migrations"),
    table_classes=table_finder(
        modules=[Table1, Table2], exclude_imported=True
    ),
    migration_dependencies=[],
    commands=[],
)
for all the apps 
```

**Reply by Kokoserver** [reply]
you can give me an example of what you meant

**Reply by dantownsend** [reply]
Sorry, I wasn't very clear. Just this:

```python
async def run():
    from piccolo_conf import DB

    for node in DB.extra_nodes:
        DB.config = node.config        
        await run_forwards("all")
```

**Reply by Kokoserver** [reply]
i have just done that, i think i'm getting something wrong it raises this error, when i reference another app tables using the lazyTableReference. 
let me show an example of what i mean:
```python 

class Designation(TenantBasedTable, tablename="ex_designation"):
    name = col.Varchar(length=100)
    employees = col.M2M(
        col.LazyTableReference(
            "DesignationToEmployee",
            app_name="employee",
        )
    )
    salary_components = col.M2M(
        col.LazyTableReference("SalaryToDesignation", app_name="benefit")
    )
    leave_types = col.M2M(
        col.LazyTableReference("LeaveToDesignation", app_name="leave")
    )


class DesignationToEmployee(TenantBasedTable, tablename="designation_to_employee"):
    designation = col.ForeignKey(
        col.LazyTableReference("Designation", app_name="designation")
    )
    employee = col.ForeignKey(col.LazyTableReference("Employee", app_name="employee"))
```

**Reply by dantownsend** [reply]
I would get rid of as many of the `LazyTableReference` as you can - in some situations it can still lead to circular imports. For example, just use `ForeignKey(Designation)`.

**Reply by Kokoserver** [reply]
alright... let me make some changes. Thanks
